test-tools/src/main/java/org/terracotta/utilities/test/logging/ConsoleAppenderCapture.java

```diff
@@ -0,0 +1,235 @@
+
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.joining;
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.Collectors.toMap;
+
+/**
+ * Captures output associated with the current thread logged to each active {@link ConsoleAppender}
+ * instance in the current logging environment.
+ */
+public class ConsoleAppenderCapture implements AutoCloseable {
+  private static final String MDC_KEY_ROOT = ConsoleAppenderCapture.class.getSimpleName() + ".testId";
+
+  private final String testId;
+  private final String mdcKey = MDC_KEY_ROOT + '.' + UUID.randomUUID();
+  private final Runnable removeAppenders;
+  private final Map<String, ListAppender<ILoggingEvent>> appenderMap = new HashMap<>();
+
+  /**
+   * Creates a new {@code ConsoleAppenderCapture} instance for {@code testId}.
+   * <p>
+   * The returned {@code ConsoleAppenderCapture} instance <b>must</b> be closed when no longer needed.
+   * Use of a try-with-resource block is recommended: <pre>{@code
+   *   try (ConsoleAppenderCapture appenderCapture = ConsoleAppenderCapture.capture("identifier")) {
+   *     // invocation of code under test using ConsoleAppender
+   *     List<String> logCapture = appenderCapture.getMessages(ConsoleAppenderCapture.Target.STDOUT);
+   *     assertThat(logCapture, Matcher...);
+   *   }
+   * }</pre>
+   *
+   * @param testId the identifier for which events are captured
+   * @return a new {@code ConsoleAppenderCapture} instance
+   * @throws IllegalStateException if the current logging configuration does not append to {@link ConsoleAppender}
+   * @see ConsoleAppenderCapture#ConsoleAppenderCapture(String)  ConsoleAppenderCapture
+   */
+  public static ConsoleAppenderCapture capture(String testId) {
+    ConsoleAppenderCapture cac = new ConsoleAppenderCapture(testId);
+    if (cac.appenderMap.isEmpty()) {
+      cac.close();
+      throw new IllegalStateException("Current logging configuration does not use " + ConsoleAppender.class.getName());
+    }
+    return cac;
+  }
+
+  /**
+   * Creates a new {@code ConsoleAppenderCapture} instance.
+   * <p>
+   * The returned {@code ConsoleAppenderCapture} instance <b>must</b> be closed when no longer needed.
+   * Use of a try-with-resource block is recommended: <pre>{@code
+   *   try (ConsoleAppenderCapture appenderCapture = ConsoleAppenderCapture.capture()) {
+   *     // invocation of code under test using ConsoleAppender
+   *     List<String> logCapture = appenderCapture.getMessages(ConsoleAppenderCapture.Target.STDOUT);
+   *     assertThat(logCapture, Matcher...);
+   *   }
+   * }</pre>
+   *
+   * @return a new {@code ConsoleAppenderCapture} instance
+   * @throws IllegalStateException if the current logging configuration does not append to {@link ConsoleAppender}
+   * @see ConsoleAppenderCapture#ConsoleAppenderCapture(String)  ConsoleAppenderCapture
+   */
+  public static ConsoleAppenderCapture capture() {
```

Comment:
It does not say if not passing a testId would then capture everything ?

Reply:
The `testId`, itself, isn't relevant to capture -- it's just for identification.  The capture is dependent on the internally-generated key which is, within the properties of `UUID`, unique to the instance of `ConsoleAppenderCapture`.  By intent, `ConsoleAppenderCapture` instances should not interfere with each other.  Hopefully, the ctor javadoc makes that clear -- if not, I'm game for re-wording.